include/deal.II/base/array_view.h

```diff
@@ -252,35 +259,97 @@ class ArrayView
    */
   const std::size_t n_elements;
 
-  friend class ArrayView<const ElementType>;
+  friend class ArrayView<const ElementType, MemorySpaceType>;
 };
 
 
 
 //---------------------------------------------------------------------------
 
 
+namespace internal
+{
+  namespace ArrayViewHelper
+  {
+    template <typename MemorySpaceType>
+    inline bool
+    correct_memory_space(const void *const)
+    {
+      return std::is_same<MemorySpaceType, MemorySpace::Host>::value;
+    }
+
+#ifdef DEAL_II_COMPILER_CUDA_AWARE
+    template <>
+    inline bool
+    correct_memory_space<MemorySpace::Host>(const void *ptr)
+    {
+      cudaPointerAttributes attributes;
+      const cudaError_t cuda_error = cudaPointerGetAttributes(&attributes, ptr);
```

Comment:
how interesting -- you can find out where the data lives by just looking at the bits of the pointer?

Reply:
Yes, `CUDA` can if there is support for Unified Virtual Addressing (UVA; from `CUDA 4` on) and the memory has been allocated by `CUDA` function (up to version 8). The current version can't distinguish between invalid memory and host memory.
From version 10 on, it can also identify host memory not managed by `CUDA` (using a different member variable).

Reply:
Correction: It is actually, `CUDA 11` where pointer to host memory will not result in an error if I read the documentation correctly.